nestkernel/random_manager.cpp

```diff
@@ -0,0 +1,228 @@
+  register_rng_type< r123::Engine< r123::Philox4x32 > >( "Philox_32" );
+#if R123_USE_PHILOX_64BIT
+  register_rng_type< r123::Engine< r123::Philox4x64 > >( "Philox_64" );
+#endif
+  register_rng_type< r123::Engine< r123::Threefry4x32 > >( "Threefry_32" );
+#if R123_USE_64BIT
+  register_rng_type< r123::Engine< r123::Threefry4x64 > >( "Threefry_64" );
+#endif
+#endif
+
+  current_rng_type_ = DEFAULT_RNG_TYPE_;
+  base_seed_ = DEFAULT_BASE_SEED_;
+
+  reset_rngs_();
+}
+
+void
+nest::RandomManager::finalize()
+{
+  for ( auto& it : rng_types_ )
+  {
+    delete it.second;
+  }
+
+  rng_types_.clear();
+  vp_specific_rngs_.clear();
+}
+
+void
+nest::RandomManager::reset_rngs_()
+{
+  // Delete existing RNGs.
+  delete rank_synced_rng_;
+
+  auto delete_rngs = []( std::vector< RngPtr >& rng_vec )
+  {
+    for ( auto rng : rng_vec )
+    {
+      delete rng;
+    }
+  };
+
+  delete_rngs( vp_synced_rngs_ );
+  delete_rngs( vp_specific_rngs_ );
+
+  // Create new RNGs of the currently used RNG type.
+  rank_synced_rng_ = rng_types_[ current_rng_type_ ]->create( { base_seed_, RANK_SYNCED_SEEDER_ } );
+
+  vp_synced_rngs_.resize( kernel().vp_manager.get_num_threads() );
+  vp_specific_rngs_.resize( kernel().vp_manager.get_num_threads() );
+
+#pragma omp parallel
+  {
+    const auto tid = kernel().vp_manager.get_thread_id();
+    const std::uint32_t vp = kernel().vp_manager.get_vp();
+    vp_synced_rngs_[ tid ] = rng_types_[ current_rng_type_ ]->create( { base_seed_, THREAD_SYNCED_SEEDER_ } );
+    vp_specific_rngs_[ tid ] = rng_types_[ current_rng_type_ ]->create( { base_seed_, THREAD_SPECIFIC_SEEDER_, vp } );
+  }
+}
+
+void
+nest::RandomManager::get_status( DictionaryDatum& d )
+{
+  ArrayDatum rng_types;
+  for ( auto rng = rng_types_.begin(); rng != rng_types_.end(); ++rng )
+  {
+    rng_types.push_back( rng->first );
+  }
+
+  def< ArrayDatum >( d, names::rng_types, rng_types );
+  def< long >( d, names::rng_seed, base_seed_ );
+  def< std::string >( d, names::rng_type, current_rng_type_ );
+}
+
+void
+nest::RandomManager::set_status( const DictionaryDatum& d )
+{
+  long rng_seed;
+  bool rng_seed_updated = updateValue< long >( d, names::rng_seed, rng_seed );
+
+  if ( rng_seed_updated )
+  {
+    if ( not( 0 < rng_seed and rng_seed < ( 1L << 32 ) ) )
+    {
+      throw BadProperty( "RNG seed must be in (0, 2^32-1)." );
+    }
+
+    base_seed_ = static_cast< std::uint32_t >( rng_seed );
+  }
+
+  std::string rng_type;
+  bool rng_type_updated = updateValue< std::string >( d, names::rng_type, rng_type );
+
+  if ( rng_type_updated )
+  {
+    auto rng = rng_types_.find( rng_type );
+    if ( rng == rng_types_.end() )
+    {
+      std::string msg = "'%1' is not a known RNG type. See /rng_types for available types";
+      throw BadProperty( String::compose( msg, rng_type ) );
+    }
+
+    current_rng_type_ = rng_type;
+  }
+
+  long n_threads;
+  bool n_threads_updated = updateValue< long >( d, names::local_num_threads, n_threads );
```

Comment:
Is the value of `n_threads` not used/set anywhere later?

Reply:
No, it's only here because it's needed in `updateValue()`. I've changed it to use `d->known()` instead.